File: backend/handler/review.go
package handler

import (
	"encoding/json"
	"log"
	"net/http"

	"github.com/pr-lens/backend/github"
)

type reviewRequest struct {
	URL      string                 `json:"url"`
	Event    string                 `json:"event"`
	Body     string                 `json:"body"`
	Comments []github.ReviewComment `json:"comments"`
}

// Review returns an http.HandlerFunc that posts a pull request review to GitHub.
// The server GITHUB_TOKEN is never used here — a client-supplied PAT is required
// so a shared backend cannot approve PRs as the bot identity.
func Review() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
			return
		}

		token, err := ResolveGitHubToken(r, "", false)
		if err != nil {
			http.Error(w, err.Error(), http.StatusUnauthorized)
			return
		}
		ghClient := github.NewClient(token)

		var req reviewRequest
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 1<<20)).Decode(&req); err != nil || req.URL == "" {
			http.Error(w, `invalid request body: expected {"url":"...","event":"..."}`, http.StatusBadRequest)
			return
		}

		switch req.Event {
		case "APPROVE", "REQUEST_CHANGES", "COMMENT":
			// valid
		default:
			http.Error(w, `event must be one of APPROVE, REQUEST_CHANGES, COMMENT`, http.StatusBadRequest)
			return
		}

		ref, err := github.ParsePRURL(req.URL)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		log.Printf("[review] posting %s review to %s/%s #%d (%d comments)", req.Event, ref.Owner, ref.Repo, ref.Number, len(req.Comments))

		if err := ghClient.PostReview(r.Context(), ref, req.Event, req.Body, req.Comments); err != nil {
			log.Printf("[review] error: %v", err)
			http.Error(w, "failed to post review: "+err.Error(), http.StatusBadGateway)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"ok":true}`))
	}
}

File: backend-ts/src/handler/review.ts
import type { Context } from "hono";
import { GitHubClient, parsePRURL } from "../github/client.js";
import { resolveGitHubToken } from "../token.js";

const VALID_EVENTS = new Set(["APPROVE", "REQUEST_CHANGES", "COMMENT"]);

export async function handleReview(c: Context, defaultToken: string): Promise<Response> {
  const token = resolveGitHubToken(c, defaultToken);
  if (!token) return c.json({ error: "GitHub token required" }, 401);

  let body: {
    url?: string;
    event?: string;
    body?: string;
    comments?: Array<{ path: string; line: number; body: string }>;
  };
  try {
    body = await c.req.json();
  } catch {
    return c.json({ error: "invalid JSON body" }, 400);
  }

  if (!body.url) return c.json({ error: "url is required" }, 400);
  if (!body.event || !VALID_EVENTS.has(body.event)) {
    return c.json({ error: "event must be one of APPROVE, REQUEST_CHANGES, COMMENT" }, 400);
  }

  let ref;
  try {
    ref = parsePRURL(body.url);
  } catch (err) {
    return c.json({ error: String(err) }, 400);
  }

  const t = Date.now();
  const commentCount = body.comments?.length ?? 0;
  console.log(`[review] start ${ref.owner}/${ref.repo} #${ref.number} ` +
    `event=${body.event} inline=${commentCount}`);

  const ghClient = new GitHubClient(token);
  try {
    await ghClient.postReview(ref, body.event, body.body ?? "", body.comments ?? []);
  } catch (err) {
    console.error(`[review] error: ${err}`);
    return c.json({ error: `failed to post review: ${err}` }, 502);
  }

  console.log(`[review] posted in ${Date.now() - t}ms`);
  return c.json({ ok: true });
}
